py/bitbox02/bitbox02/util.py

```diff
@@ -39,3 +46,90 @@ def parse_xpub(xpub: str) -> common.XPub:
+
+
+class UserCache:
+    """Data structure to hold keys"""
+
+    # pylint: disable=too-few-public-methods
+    def __init__(self, raw_cache: Optional[str] = None):
+        if raw_cache is None:
+            self.app_static_privkey = None
+            self.device_static_pubkeys: List[bytes] = []
+            return
+        (privkey, pubkeys) = UserCache.deserialize(raw_cache)
+        self.app_static_privkey = privkey
+        self.device_static_pubkeys = pubkeys
+
+    def serialize(self) -> str:
+        """Serialize struct to string"""
+        pubkeys = [binascii.hexlify(x).decode("utf-8") for x in self.device_static_pubkeys]
+        privkey = None
+        if self.app_static_privkey is not None:
+            privkey = binascii.hexlify(self.app_static_privkey).decode("utf-8")
+        return json.dumps({"device_static_pubkeys": pubkeys, "app_static_privkey": privkey})
+
+    @staticmethod
+    def deserialize(raw: str) -> Tuple[Optional[bytes], List[bytes]]:
+        """Deserialize content from disk to struct"""
+        try:
+            data = json.loads(raw)
+            privkey = None
+            if data["app_static_privkey"] is not None:
+                privkey = binascii.unhexlify(data["app_static_privkey"])
+            pubkeys = [binascii.unhexlify(x) for x in data["device_static_pubkeys"]]
+            return (privkey, pubkeys)
+        except json.JSONDecodeError:
+            return (None, [])
+
+
+class NoiseConfigUserCache(bitbox_api_protocol.BitBoxNoiseConfig):
+    """A noise config that stores the keys in a file in XDG_CACHE_HOME or ~/.cache"""
+
+    def __init__(self) -> None:
+        self._cache_file_path = NoiseConfigUserCache._find_cache_file()
+        super().__init__()
+
+    @staticmethod
+    def _find_cache_file() -> Path:
+        cachedir_env = os.environ.get("XDG_CACHE_HOME", "")
+        if cachedir_env == "":
+            homedir = os.environ.get("HOME", "")
+            if homedir == "":
+                raise RuntimeError("Can't find cache dir")
+            cachedir = Path(homedir) / ".cache"
+        else:
+            cachedir = Path(cachedir_env)
+        return cachedir / "shift" / "bitbox02.dat"
+
+    def _read_cache(self) -> UserCache:
+        try:
+            with self._cache_file_path.open("r") as fileh:
+                return UserCache(fileh.read())
+        except FileNotFoundError:
+            return UserCache()
+
+    def _write_cache(self, data: UserCache) -> None:
+        self._cache_file_path.parent.mkdir(parents=True, exist_ok=True)
+        with self._cache_file_path.open("w") as fileh:
+            fileh.write(data.serialize())
+
+    def contains_device_static_pubkey(self, pubkey: bytes) -> bool:
+        data = self._read_cache()
+        if pubkey in data.device_static_pubkeys:
+            return True
+        return False
+
+    def add_device_static_pubkey(self, pubkey: bytes) -> None:
+        data = self._read_cache()
+        data.device_static_pubkeys.append(pubkey)
```

Comment:
this also needs to check if it's already in there (i guess the four methods should have been documented ;) )

Reply:
Ack